src/Dialog/DialogBottomSheet.tsx

```diff
@@ -0,0 +1,327 @@
+import React, {useEffect, useRef, useState, PropsWithChildren, MouseEvent, TouchEvent, ChangeEvent} from 'react'
+import styled from 'styled-components'
+import {SxProp} from '../sx'
+import {useRefObjectAsForwardedRef} from '../hooks/useRefObjectAsForwardedRef'
+import Box from '../Box'
+import {get} from '../constants'
+
+const ANIMATION_DURATION = 300
+const FULL_HEIGHT = 95
+const HALF_HEIGHT = 50
+
+/**
+ * Props to customize the rendering of the Dialog.
+ */
+export interface DialogBottomSheetProps extends SxProp {
+  /**
+   * This method is invoked when a gesture to close the dialog is used
+   */
+  onClose: (gesture: 'close-button' | 'escape' | 'drag' | 'overlay') => void
+
+  /**
+   * Passed through to use as the draggable area of the dialog.
+   */
+  header: React.ReactNode
+
+  /**
+   * Default: "dialog". The ARIA role to assign to this dialog.
+   * @see https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal
+   * @see https://www.w3.org/TR/wai-aria-practices-1.1/#alertdialog
+   */
+  role?: 'dialog' | 'alertdialog'
+
+  /**
+   * Identifies the element (or elements) that labels the element it is applied to.
+   */
+  ariaLabelledby: string
+
+  /**
+   * Identifies the element (or elements) that describes the element on which the attribute is set.
+   */
+  ariaDescribedby: string
+
+  /**
+   * Indicates whether an element is modal when displayed.
+   */
+  ariaModal: boolean
+}
+
+export default React.forwardRef<HTMLDivElement, PropsWithChildren<DialogBottomSheetProps>>((props, forwardedRef) => {
+  const {onClose, children, role, ariaLabelledby, ariaDescribedby, ariaModal, header, sx} = props
+
+  // 🔄 STATES
+
+  const [open, setIsOpen] = useState<boolean>(false)
+  const [snappedHeight, setSnappedHeight] = useState<number>(HALF_HEIGHT)
+  const [fireDelayedOnClose, setFireDelayedOnClose] = useState<
+    'close-button' | 'escape' | 'drag' | 'overlay' | undefined
+  >()
+
+  // 📎 REFERENCES
+
+  const dialogRef = useRef<HTMLDivElement>(null)
+  const startY = useRef(0)
+  const startHeight = useRef(0)
+  const isDragging = useRef(false)
+
+  useRefObjectAsForwardedRef(forwardedRef, dialogRef)
+
+  // 🧑‍🦽 ACCESSIBILITY
+
+  const isReduced = prefersReducedMotion()
+
+  // 🥊 ACTIONS
+
+  const showBottomSheet = () => {
+    setIsOpen(true)
+    updateSheetHeight(HALF_HEIGHT)
+  }
+
+  const hideBottomSheet = async (gesture: 'close-button' | 'escape' | 'drag' | 'overlay') => {
+    setIsOpen(false)
+    if (isReduced) {
+      onClose(gesture)
+    } else {
+      setFireDelayedOnClose(gesture)
+    }
+  }
+
+  const updateSheetHeight = (height: number) => {
+    if (!dialogRef.current) return
+    dialogRef.current.style.height = `${height}dvh`
+    setSnappedHeight(height)
```

Comment:
We might get better performance if we always render it at full height and use `translate: transformY();` to position it in and out of the viewport.

Changing `height` triggers reflow (and maybe repaint?).

Reply:
Are you sure it will reflow/repaint because the content is scrollable therefore the width and height of the scrollable content won't change. I haven't observed any performance issues. Not animating the height is kinda complex because the scrollable area needs to resize either way. Using `transform` I would have to set a height before animating to something beyond the height of the viewport and at the end of the animation then set the final height to ensure the scrollable area is correct. Things get even more tricky while you're dragging.

Reply:
I checked it out in Chrome's performance dev tools, and it seems to behave just fine on my work computer. Let's leave this as-is unless we notice jank on mobile or other less powerful devices. 🙂 

---

> Are you sure it will reflow/repaint because the content is scrollable therefore the width and height of the scrollable content won't change.

I'm pretty sure the browser does a reflow whenever you change a height or width, even if no content is actually affected. Browsers even trigger a reflow when you check the dimensions of an element using `.getBoundingClientRect()` 😢

Reply:
Oh right. I've learned something today. 🧠